File: src/leetcode/first/Code_227_BasicCalculator2.java
package leetcode.first;

import java.util.Stack;

/**
 * 基本计算器
 * 没有括号的加减乘除四则运算
 *
 * @since 2020-6-16 Thursday 17:26
 */
public class Code_227_BasicCalculator2 {
    static int calculate(String s) {
        int res = 0, d = 0;
        char sign = '+';
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) >= '0') {
                d = d * 10 + (s.charAt(i) - '0');
            }
            if ((s.charAt(i) < '0' && s.charAt(i) != ' ') || i == s.length() - 1) {
                switch (sign) {
                    case '+':
                        stack.push(d);
                        break;
                    case '-':
                        stack.push(-d);
                        break;
                    case '*':
                        int m = stack.pop();
                        stack.push(d * m);
                        break;
                    case '/':
                        int n = stack.pop();
                        stack.push(n / d);
                        break;
                    default:
                        break;
                }
                sign = s.charAt(i);
                d = 0;
            }
        }
        for (Integer i : stack) {
            res += i;
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println(calculate("1 + 2 * 3"));
        System.out.println(calculate("1 * 2 - 3"));
    }
}

File: src/cpp-leetcode/first/Code_227_BasicCalculator2.cpp
#include <iostream>
#include <stack>
#include <string>

using std::string;

/**
 * 不包含括号的加减乘除四则运算
 *
 * @since 2020-10-27 Tuesday 16:30 - 17:12
 */
class Solution {
   public:
    int calculate(string s) {
        int size = s.size();
        if (size == 0) {
            return 0;
        }
        std::stack<int> numbers;
        int res = 0, num = 0;
        char sign = '+';
        for (int i = 0; i < size; i++) {
            char c = s[i];
            if (c >= '0' && c <= '9') {
                num = num * 10 + (c - '0');
            }
            if (((c < '0' || c > '9') && c != ' ') || i == size - 1) {
                switch (sign) {
                    case '+':
                        numbers.push(num);
                        break;
                    case '-':
                        numbers.push(-num);
                        break;
                    case '*': {
                        int m = numbers.top();
                        numbers.pop();
                        numbers.push(m * num);
                        break;
                    }
                    case '/': {
                        int m = numbers.top();
                        numbers.pop();
                        numbers.push(m / num);
                        break;
                    }
                    default:
                        break;
                }
                num = 0;
                sign = c;
            }
        }
        while (!numbers.empty()) {
            res += numbers.top();
            numbers.pop();
        }
        return res;
    }
};

int main(int argc, char const *argv[]) {
    Solution s;
    string s1 = "3+2*2";
    string s2 = " 3/2 ";
    string s3 = " 3+5 / 2 ";
    string s4 = "1-1+1";
    std::cout << s.calculate(s1) << "\n";  // 7
    std::cout << s.calculate(s2) << "\n";  // 1
    std::cout << s.calculate(s3) << "\n";  // 5
    std::cout << s.calculate(s4) << "\n";  // 1
}
